### PR TITLE
shell: Add shell_exec and shell tests

### PR DESCRIPTION
**Pull request by Vudentz**
This adds shell_exec which can be used to execute commands directly
without the use of a console which is useful for both testing as well
as interfacing with applications/upper layer which would like to have
access to shell commands directly.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

### CODE REVIEW

**Review by jukkar — COMMENTED** [review]
@pfalcon Would it make sense in micropython to use this instead of calling the net-shell C-functions directly?

**Review by jukkar — DISMISSED** [review]
I like this approach so +1 from me.

**Comment by jhedberg** [comment]
A couple of comments:

Make it super clear in the documentation that the shell commands are not a formal API or something that can be considered stable. I.e. anyone writing code that uses the command strings should be prepared for that code to stop working at some point.

Secondly, make it clear in the documentation that C string literals cannot be passed to this function because it modifies the input (by inserting nul characters to terminate each command line argument). Instead, the caller will need to do something like "char cmd[] = "my command"; shell_exec(cmd);"

**Review by nashif — CHANGES_REQUESTED** [review]
Please provide a testcase for this new usage and document this in the shell subsystem documentation.

**Review by jukkar — DISMISSED** [review]
LGTM

**Comment by Vudentz** [comment]
@nashif Tests are now included.

File: tests/shell/testcase.ini [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[test]
+build_only = true
```
Comment (nashif) [inline tests/shell/testcase.ini]: remove, this should be run in qemu during sanitycheck

File: tests/shell/src/main.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
+/** @file
+ *  @brief Interactive shell test suite
+ *
+ */
+
```
Comment (nashif) [inline tests/shell/src/main.c]: put this below the license header

File: tests/shell/src/main.c [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	ret = shell_exec(cmd);
+
+	TC_PRINT("shell_exec(%s): %d\n", line, ret);
+
+	zassert_true(ret == result, line);
+}
+
+static void test_help(void)
+{
+	TC_PRINT("\nhelp tests:\n");
```
Comment (nashif) [inline tests/shell/src/main.c]: this duplicates what the testsuite already prints, too verbose. I would remove all those prints, thats how it looks now:

`tc_start() - test_select

select tests:`

File: tests/shell/src/main.c [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+	test_shell_exec("help", 0);
+	test_shell_exec("help dummy", 0);
+	test_shell_exec("help invalid", -EINVAL);
+}
+
+static void test_select(void)
+{
+	TC_PRINT("\nselect tests:\n");
```
Comment (nashif) [inline tests/shell/src/main.c]: same, remove this, this is already printed by ztest, other tests below as well.

**Review by nashif — CHANGES_REQUESTED** [review]
Please move to test to:

tests/subsys/shell

**Review by pfalcon — APPROVED** [review]
>@pfalcon Would it make sense in micropython to use this instead of calling the net-shell C-functions directly?

As discussed in the previous patch by @Vudentz, using IP stack diagnostic functions is a more direct way. We don't have such functions well defined, but un-static-izing functions used by net shell was a first step in that direction.

These changes are definitely good improvements to the shell system, and yep, if direct access to net subsystem functions will be closed, can serve as a (somewhat hackish) fallback for that too.

File: tests/shell/testcase.ini (hunk 1 above)
Comment (Vudentz) [inline tests/shell/testcase.ini]: Done

File: tests/shell/src/main.c (hunk 2 above)
Comment (Vudentz) [inline tests/shell/src/main.c]: Done

File: tests/shell/src/main.c (hunk 3 above)
Comment (Vudentz) [inline tests/shell/src/main.c]: Done

File: tests/shell/src/main.c (hunk 4 above)
Comment (Vudentz) [inline tests/shell/src/main.c]: Done

**Comment by Vudentz** [comment]
> As discussed in the previous patch by @Vudentz, using IP stack diagnostic functions is a more direct way. We don't have such functions well defined, but un-static-izing functions used by net shell was a first step in that direction.

> These changes are definitely good improvements to the shell system, and yep, if direct access to net subsystem functions will be closed, can serve as a (somewhat hackish) fallback for that too.

Im not directly targeting any changes to net_shell.h right now, though this offer an alternative to having the callback handler exposed in headers like that I guess we need a proper discussion on what are the benefit vs shell_exec.

**Review by nashif — APPROVED** [review]

**Pull request merged**